source/extensions/http/header_validators/envoy_default/path_normalizer.cc

```diff
@@ -0,0 +1,348 @@
+      // default implementation: normalize the encoded octet and accept the path
+      return {PercentDecodeResult::Normalized};
+
+    case HeaderValidatorConfig_UriPathNormalizationOptions::REJECT_REQUEST:
+      // Reject the entire request
+      return {PercentDecodeResult::Reject};
+
+    case HeaderValidatorConfig_UriPathNormalizationOptions::UNESCAPE_AND_FORWARD:
+      // Decode the slash and accept the path.
+      return {PercentDecodeResult::Decoded, ch};
+
+    case HeaderValidatorConfig_UriPathNormalizationOptions::UNESCAPE_AND_REDIRECT:
+      // Decode the slash and response with a redirect to the normalized path.
+      return {PercentDecodeResult::DecodedRedirect, ch};
+
+    default:
+      // This should never occur but it's here to make the compiler happy because of the extra
+      // values added by protobuf.
+      break;
+    }
+  }
+
+  // The octet is a valid encoding but it wasn't be decoded because it was outside the UNRESERVED
+  // character set.
+  return {PercentDecodeResult::Normalized};
+}
+
+/*
+ * Find the start index of the previous segment within the path. The previous segment starts at the
+ * first non-slash character after the preceding slash. For example:
+ *
+ *   path = "/hello/world/..";
+ *                  ^    ^-- iterator
+ *                  |-- start of previous segment
+ *
+ * The ``begin`` iterator is returned on error.
+ */
+std::string::iterator findStartOfPreviousSegment(std::string::iterator iter,
+                                                 std::string::iterator begin) {
+  bool seen_segment_char = false;
+  for (; iter != begin; --iter) {
+    if (*iter == '/' && seen_segment_char) {
+      ++iter;
+      return iter;
+    }
+
+    if (*iter != '/' && !seen_segment_char) {
+      seen_segment_char = true;
+    }
+  }
+
+  if (seen_segment_char) {
+    ++begin;
+  }
+
+  return begin;
+}
+
+PathNormalizer::PathNormalizationResult
+PathNormalizer::normalizePathUri(RequestHeaderMap& header_map) const {
+  // Attempt to parse the request-target to see if it is in origin-form. We are only normalizing
+  // the path component of the request-target, so if is is origin or absolute form we will extract
+  // the path. From RFC 9112, https://www.rfc-editor.org/rfc/rfc9112.html#section-3.2:
+  //
+  // request-target = origin-form
+  //                / absolute-form
+  //                / authority-form
+  //                / asterisk-form
+  //
+  // origin-form    = absolute-path [ "?" query ]
+  // absolute-form  = absolute-URI
+  // authority-form = uri-host ":" port
+  // asterisk-form  = "*"
+  const auto original_uri = header_map.path();
+  if (original_uri == "*") {
+    // Asterisk form
+    return PathNormalizationResult::success();
+  }
+
+  const bool is_connect_method =
+      header_map.method() == ::Envoy::Http::Headers::get().MethodValues.Connect;
+  ::Envoy::Http::Utility::Url url;
+  const bool is_origin_form = !url.initialize(original_uri, is_connect_method);
```

Comment:
This will cause the URL to be parsed twice in H/1 case. Once here and then another one in the validateRequestHeaderMap when request target is in the absolute form. I think we should take care of the absolute form  request target before invoking normalizePathUri to avoid parsing twice.

Reply:
We still need the ability to pull out the `path` component of the `:path` header. So I added a couple methods that attempt to split `:path` into scheme/authority, path, and query/fragment. Then, after normalization has finished, only the path component is updated in the header, preserving the scheme/authority and query/fragment.